File: client/src/lib/mal.ts
/**
 * MyAnimeList link and sync-log data layer (spec §4.7 FR-M1–FR-M7, §5,
 * roadmap M9).
 *
 * Two things live here. The **link** — an OAuth grant the server holds on the
 * viewer's behalf — which the client only ever starts, inspects, or tears
 * down; and the **write log**, the record of every change Arc pushed to MAL
 * with the value it replaced (FR-M5). The log is the reason FR-M7 is
 * checkable from the outside: if Arc wrote it, it is in this list, with the
 * user event that caused it.
 *
 * Starting a link is a full-page navigation, not a fetch: the server answers
 * with MAL's authorize URL and the browser has to *go there* so the person can
 * sign in on MAL's own origin. That is the one place in the client where a
 * mutation ends by leaving the app.
 *
 * Everything an import, a push or a revert does lands in two caches — the MAL
 * views here and the show pages, whose `list_entry.mal_sync` badge is derived
 * from the same writes — so every mutation invalidates both.
 */

import {
  useMutation,
  useQuery,
  useQueryClient,
  type UseMutationResult,
  type UseQueryResult,
} from '@tanstack/react-query'
import { ApiError, apiFetch } from '@/lib/api'
import { ANIME_QUERY_KEY, isListStatus, LIST_STATUS_LABELS, type AnimeSummary } from '@/lib/anime'
import { errorDetail } from '@/lib/auth'

/** `GET /api/mal/status` — the state of this viewer's MAL link. */
export interface MalStatus {
  linked: boolean
  /** The MAL account Arc is linked to; null when there is no link. */
  mal_username: string | null
  /** When the current access token expires; null when there is no link. */
  expires_at: string | null
  /** Last successful import of the MAL list (FR-M2, FR-M3). */
  last_import_at: string | null
  /** The refresh token no longer works: the person has to authorise again. */
  needs_relink: boolean
  /**
   * Fields queued but not yet accepted by MAL (FR-M6). One per queued row, so
   * a single show whose status, score and progress all moved counts three.
   */
  pending_writes: number
  /** Writes that exhausted their retries and need attention (FR-M6). */
  failed_writes: number
  /** False when the server has no MAL client id, so linking cannot be offered. */
  configured: boolean
}

/** The three fields Arc is allowed to write to MAL (FR-M4). */
export type MalWriteField = 'status' | 'score' | 'progress'

/**
 * What caused a write. Every one of these is user-originated (FR-M7):
 * `watch` is an episode completed, `manual` an explicit edit in Arc, `revert`
 * an entry undone from this log, `conflict` the losing side of a two-sided
 * change being written back (FR-M3).
 */
export type MalWriteCause = 'watch' | 'manual' | 'revert' | 'conflict'

/**
 * Where a write got to. `pending` is written before the request goes out, so a
 * crash mid-write leaves evidence rather than silence; it becomes `ok` or
 * `failed` when MAL answers. `skipped` means nothing was sent at all — the
 * FR-M4 rule declining to lower progress, a change overtaken by a later one, a
 * value MAL already had.
 */
export type MalWriteStatus = 'pending' | 'ok' | 'failed' | 'skipped'

/** One row of `GET /api/mal/log` — a write, and the value it replaced (FR-M5). */
export interface MalWrite {
  id: number
  /** Null when the show has since left the catalogue; the row still stands. */
  anime: AnimeSummary | null
  field: MalWriteField
  /** The value before the write; `formatMalValue` turns it into words. */
  old_value: unknown
  new_value: unknown
  cause: MalWriteCause
  status: MalWriteStatus
  /**
   * MAL's own complaint, for a `failed` row — and for a `skipped` one, always
   * a sentence saying why nothing was sent ("automatic progress never lowers
   * MAL"), which is an explanation rather than a fault.
   */
  error: string | null
  created_at: string
  /** False once a row has been reverted, or when there is nothing to go back to. */
  revertible: boolean
}

/** Per-show sync state, carried on `AnimeDetail.list_entry` (FR-M6). */
export interface MalSync {
  state: 'synced' | 'pending' | 'failed' | 'unlinked'
  error: string | null
  last_write_at: string | null
}

/** Which rows the log page is asking for. `status` unset means every row. */
export interface MalLogFilters {
  limit?: number
  status?: MalWriteStatus
}

export const MAL_QUERY_KEY = 'mal'

export const malStatusQueryKey = [MAL_QUERY_KEY, 'status'] as const

export function malLogQueryKey(
  filters: MalLogFilters = {},
): readonly [string, string, string, number | 'default'] {
  return [MAL_QUERY_KEY, 'log', filters.status ?? 'all', filters.limit ?? 'default']
}

/** How many rows the log asks for when the caller does not say. */
export const MAL_LOG_LIMIT = 50

/** What a null / absent value reads as, in a table cell or a value pair. */
export const EMPTY_VALUE = '—'

export const MAL_FIELD_LABELS: Record<MalWriteField, string> = {
  status: 'Status',
  score: 'Score',
  progress: 'Progress',
}

export const MAL_CAUSE_LABELS: Record<MalWriteCause, string> = {
  watch: 'Watched',
  manual: 'Manual change',
  revert: 'Revert',
  conflict: 'Conflict',
}

export const MAL_WRITE_STATUS_LABELS: Record<MalWriteStatus, string> = {
  pending: 'Pending',
  ok: 'Synced',
  failed: 'Failed',
  skipped: 'Skipped',
}

const OK_BADGE = 'text-[var(--arc-ok)] border-[var(--arc-ok)]/40 bg-[var(--arc-ok)]/10'
const BUSY_BADGE =
  'text-[var(--arc-accent)] border-[var(--arc-accent)]/40 bg-[var(--arc-accent)]/10'
const BAD_BADGE = 'text-[var(--arc-error)] border-[var(--arc-error)]/40 bg-[var(--arc-error)]/10'
const MUTED_BADGE =
  'text-[var(--arc-text-muted)] border-[var(--arc-border)] bg-[var(--arc-surface-raised)]'

export const MAL_WRITE_STATUS_CLASSES: Record<MalWriteStatus, string> = {
  ok: OK_BADGE,
  pending: BUSY_BADGE,
  failed: BAD_BADGE,
  // Skipped is not a failure — the FR-M4 rule working — so it reads as quiet.
  skipped: MUTED_BADGE,
}

/**
 * A value as a person reads it. A status comes back as the wire enum
 * (`on_hold`), which is not what the rest of the client shows anywhere else;
 * a score or a progress count is already a number. Everything absent — a
 * first-ever write has no previous value — is the same em dash the score
 * control uses for "unset".
 */
export function formatMalValue(field: MalWriteField, value: unknown): string {
  if (value === null || value === undefined) return EMPTY_VALUE
  if (typeof value === 'number') return String(value)
  if (typeof value === 'string') {
    if (value === '') return EMPTY_VALUE
    if (field === 'status' && isListStatus(value)) return LIST_STATUS_LABELS[value]
    return value
  }
  if (typeof value === 'boolean') return value ? 'Yes' : 'No'
  // Nothing the server sends should land here; showing the JSON beats a blank
  // cell if it ever does.
  return JSON.stringify(value)
}

/** A timestamp as a date and a time; the log is read in minutes, not days. */
export function formatMalDate(iso: string | null | undefined): string {
  if (iso === null || iso === undefined || iso === '') return EMPTY_VALUE
  const at = new Date(iso)
  if (Number.isNaN(at.getTime())) return EMPTY_VALUE
  return new Intl.DateTimeFormat(undefined, {
    day: 'numeric',
    month: 'short',
    year: 'numeric',
    hour: 'numeric',
    minute: '2-digit',
  }).format(at)
}

/**
 * The `error` codes MAL (or Arc's callback) can send back on the redirect.
 * Anything not listed is shown with its code, which is at least something to
 * quote in a bug report.
 */
const MAL_LINK_ERRORS: Record<string, string> = {
  access_denied: 'You declined the request on MyAnimeList, so nothing was linked.',
  invalid_state: 'That link attempt expired or was tampered with. Try connecting again.',
  state_mismatch: 'That link attempt expired or was tampered with. Try connecting again.',
  missing_code: 'MyAnimeList did not send an authorisation code back. Try connecting again.',
  token_exchange_failed: 'MyAnimeList refused to issue a token. Try connecting again.',
  not_configured: 'This server has no MyAnimeList client id, so linking is off.',
  server_error: 'MyAnimeList had a problem finishing the link. Try again in a few minutes.',
}

export function malLinkErrorMessage(code: string): string {
  return MAL_LINK_ERRORS[code] ?? `MyAnimeList could not be linked (${code}).`
}

/** What to say when a MAL call fails; the log page shows this verbatim. */
export function malErrorMessage(error: unknown): string {
  if (!(error instanceof ApiError)) return 'Could not reach the server. Try again.'
  switch (error.status) {
    case 401:
      return 'Your session expired. Sign in again.'
    case 404:
      return 'That entry is gone.'
    case 409:
      return errorDetail(error) ?? 'That is no longer possible — reload the page.'
    case 503:
      return 'This server has no MyAnimeList client id, so linking is off.'
    default:
      return 'Something went wrong. Try again.'
  }
}

/** The link state for this viewer. Cheap, and every action here changes it. */
export function useMalStatus(): UseQueryResult<MalStatus, Error> {
  return useQuery<MalStatus, Error>({
    queryKey: malStatusQueryKey,
    queryFn: () => apiFetch<MalStatus>('/api/mal/status'),
    retry: false,
  })
}

export function useMalLog(filters: MalLogFilters = {}): UseQueryResult<MalWrite[], Error> {
  const limit = filters.limit ?? MAL_LOG_LIMIT
  const status = filters.status

  return useQuery<MalWrite[], Error>({
    queryKey: malLogQueryKey({ limit, status }),
    queryFn: () => {
      const params = new URLSearchParams({ limit: String(limit) })
      if (status !== undefined) params.set('status', status)
      return apiFetch<MalWrite[]>(`/api/mal/log?${params.toString()}`)
    },
    retry: false,
  })
}

/** `POST /api/mal/link`: where MAL wants the person to sign in. */
interface MalLinkStart {
  authorize_url: string
}

/**
 * Start the OAuth flow (FR-M1). The server mints the PKCE challenge and the
 * state, so all the client does with the answer is *go there* — a fetch of the
 * authorize URL would land in a login page the app cannot render and would
 * lose the redirect back to `/mal`.
 *
 * There is deliberately no cache work on success: the page is about to be
 * replaced by MAL's.
 */
export function useStartMalLink(): UseMutationResult<MalLinkStart, Error, void> {
  return useMutation<MalLinkStart, Error, void>({
    mutationFn: () => apiFetch<MalLinkStart>('/api/mal/link', { method: 'POST' }),
    onSuccess: ({ authorize_url }) => {
      window.location.assign(authorize_url)
    },
  })
}

/**
 * Every MAL action changes both the link state and what a show page says about
 * its own sync (`list_entry.mal_sync`), and all of it is server-derived — a
 * queued push, an import that rewrote list entries — so there is nothing
 * honest to patch a cache to. Both trees are invalidated and re-asked.
 */
function useMalMutation<TResult>(
  request: () => Promise<TResult>,
): UseMutationResult<TResult, Error, void> {
  const client = useQueryClient()

  return useMutation<TResult, Error, void>({
    mutationFn: request,
    onSuccess: () => {
      void client.invalidateQueries({ queryKey: [MAL_QUERY_KEY] })
      void client.invalidateQueries({ queryKey: [ANIME_QUERY_KEY] })
    },
  })
}

/** Drop the link and the stored tokens (FR-M1). Arc stops writing at once. */
export function useUnlinkMal(): UseMutationResult<null, Error, void> {
  return useMalMutation<null>(() => apiFetch<null>('/api/mal/link', { method: 'DELETE' }))
}

/** Re-import the MAL list as the baseline (FR-M2). 202: a job does the work. */
export function useMalImport(): UseMutationResult<null, Error, void> {
  return useMalMutation<null>(() => apiFetch<null>('/api/mal/import', { method: 'POST' }))
}

/** Retry the queued writes now instead of waiting for the backoff (FR-M6). */
export function useMalPush(): UseMutationResult<null, Error, void> {
  return useMalMutation<null>(() => apiFetch<null>('/api/mal/push', { method: 'POST' }))
}

/**
 * Put a logged write's previous value back (FR-M5). This is itself a
 * user-originated write, so it is queued and logged like any other — the log
 * grows a `revert` row rather than the original one disappearing.
 */
export function useRevertMalWrite(): UseMutationResult<null, Error, number> {
  const client = useQueryClient()

  return useMutation<null, Error, number>({
    mutationFn: (id) => apiFetch<null>(`/api/mal/log/${String(id)}/revert`, { method: 'POST' }),
    onSuccess: () => {
      void client.invalidateQueries({ queryKey: [MAL_QUERY_KEY] })
      void client.invalidateQueries({ queryKey: [ANIME_QUERY_KEY] })
    },
  })
}

File: server/arc/api/mal.py
"""The MyAnimeList endpoints: link, import, push, log, revert (spec §4.7).

Every route is scoped to the caller. There is no admin view of somebody else's
MAL account and no route takes a user id — a MAL link is the most personal
thing Arc stores, and the shape of the API is where that is enforced rather
than in a check somebody could forget.

**The callback is the odd one.** ``GET /api/mal/callback`` is where the
browser lands coming back from myanimelist.net, so it is reached by a top-level
navigation rather than by the client's fetch layer. Three consequences:

* It still requires a session. ``SameSite=Lax`` sends the cookie on exactly
  this kind of navigation, so the requirement costs nothing and closes the
  hole where a link could be completed by someone who is not signed in.
* It checks the session user against the user inside the ``state``. The state
  is encrypted, which proves *Arc* issued it, not that this browser is the one
  it was issued to; without this check a state pasted into a second browser
  would attach a MyAnimeList account to the wrong Arc user.
* It answers with a redirect rather than JSON, because a person is looking at
  it. Success and every kind of failure land on the same client page with a
  query parameter, so the client has one place to render both.

Nothing in this module writes to MyAnimeList. The link route hands back a URL,
the callback stores tokens, and everything else queues a job — which is what
keeps FR-M7's "one enforcement point" true: the revert below is the third and
last caller of :func:`~arc.services.mal.names.enqueue_mal_push`.
"""

from __future__ import annotations

import logging
from datetime import UTC, datetime
from typing import Annotated, Any
from urllib.parse import urlencode

from fastapi import APIRouter, HTTPException, Query, Response, status
from fastapi.responses import RedirectResponse
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from arc.api.anime_schemas import AnimeSummary, MalSyncOut
from arc.api.deps import CurrentUser, SessionDep, SettingsDep
from arc.config import ConfigurationError, Settings
from arc.core.crypto import InvalidToken
from arc.models import (
    Anime,
    ListEntry,
    ListStatus,
    MalLink,
    MalWriteCause,
    MalWriteLog,
    MalWriteStatus,
    UpdatedBy,
)
from arc.services.mal import oauth, writelog
from arc.services.mal.client import MalApiError, needs_relink, store_tokens
from arc.services.mal.factory import client_of, oauth_client
from arc.services.mal.names import (
    enqueue_mal_import,
    enqueue_mal_push,
    enqueue_mal_push_all,
)
from arc.services.mal.sync import SKIP_DISCONNECTED, apply_change, discard_pending

log = logging.getLogger(__name__)

router = APIRouter(prefix="/api/mal", tags=["mal"])

NOT_CONFIGURED = "MyAnimeList is not configured on this server"
ALREADY_LINKED = "this account is already linked to MyAnimeList"
NOT_LINKED = "this account is not linked to MyAnimeList"
NOT_FOUND = "no such write log entry"
NOT_REVERTIBLE = "this write has been superseded and can no longer be reverted"
NEWER_QUEUED = "a newer change to this field is still queued"

#: Where the client's MAL page lives, relative to ``PUBLIC_URL``. The callback
#: redirects here with either ``?linked=1`` or ``?error=<code>``.
CLIENT_PATH = "/mal"

#: Error codes the callback can put in that query string. Short, stable and
#: machine-readable: the client turns them into a sentence, so changing the
#: wording never means changing this file.
ERROR_STATE = "invalid_state"
ERROR_EXCHANGE = "exchange_failed"

#: Default and maximum rows returned by the log endpoint.
LOG_LIMIT = 50
LOG_MAX_LIMIT = 200


class MalStatusOut(BaseModel):
    """``GET /api/mal/status`` — everything the MAL page's header needs."""

    #: Whether this server has a ``MAL_CLIENT_ID`` at all. False means the
    #: link button is pointless and the client says so rather than offering it.
    configured: bool
    linked: bool
    mal_username: str | None = None
    #: When the current access token lapses. Null while the link needs
    #: re-authorising, since there is no token left to expire.
    expires_at: datetime | None = None
    last_import_at: datetime | None = None
    #: True when MyAnimeList rejected the stored refresh token: the row is
    #: still there, its credentials are not, and only the user can fix it.
    needs_relink: bool = False
    #: Changes Arc still owes MyAnimeList, and fields whose last attempt
    #: failed (:mod:`arc.services.mal.writelog`).
    pending_writes: int = 0
    failed_writes: int = 0


class MalLinkOut(BaseModel):
    """``POST /api/mal/link`` — where to send the browser."""

    authorize_url: str


class MalWriteOut(BaseModel):
    """One row of the user's write log (FR-M5)."""

    id: int
    #: Null only if the cached show has been purged out from under the log.
    anime: AnimeSummary | None = None
    field: str
    old_value: Any = None
    new_value: Any = None
    cause: MalWriteCause
    status: MalWriteStatus
    error: str | None = None
    created_at: datetime
    #: Whether ``POST /api/mal/log/{id}/revert`` would be accepted: this write
    #: succeeded and nothing has overwritten its field since.
    revertible: bool = False


class RevertOut(BaseModel):
    """``POST /api/mal/log/{id}/revert`` — what was queued."""

    anime_id: int
    field: str
    value: Any = None
    job_id: int | None = None


async def _link_of(session: AsyncSession, user_id: int) -> MalLink | None:
    return await session.get(MalLink, user_id)


@router.get("/status", response_model=MalStatusOut, summary="This user's MAL link state")
async def link_status(
    user: CurrentUser, session: SessionDep, settings: SettingsDep
) -> MalStatusOut:
    link = await _link_of(session, user.id)
    return MalStatusOut(
        configured=bool(settings.mal_client_id),
        linked=link is not None,
        mal_username=link.mal_username if link else None,
        expires_at=link.expires_at if link else None,
        last_import_at=link.last_import_at if link else None,
        needs_relink=needs_relink(link) if link else False,
        pending_writes=await writelog.pending_count(session, user_id=user.id),
        failed_writes=await writelog.failed_count(session, user_id=user.id),
    )


@router.post(
    "/link",
    response_model=MalLinkOut,
    summary="Begin the MyAnimeList OAuth handshake (FR-M1)",
    responses={
        409: {"description": ALREADY_LINKED},
        503: {"description": NOT_CONFIGURED},
    },
)
async def begin_link(user: CurrentUser, session: SessionDep, settings: SettingsDep) -> MalLinkOut:
    """Return the URL to send the browser to.

    Nothing is stored: the verifier and the caller's id travel inside the
    encrypted ``state`` (:mod:`arc.services.mal.oauth`), so an abandoned link
    attempt leaves no row to expire.
    """
    if not settings.mal_client_id:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=NOT_CONFIGURED)
    existing = await _link_of(session, user.id)
    if existing is not None and not needs_relink(existing):
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=ALREADY_LINKED)

    verifier = oauth.new_verifier()
    try:
        state = oauth.encode_state(
            settings, user_id=user.id, verifier=verifier, now=datetime.now(UTC)
        )
        return MalLinkOut(
            authorize_url=oauth.authorize_url(settings, state=state, verifier=verifier)
        )
    except ConfigurationError as exc:
        # No FERNET_KEY: the state cannot be sealed, so the handshake cannot
        # start. Same answer as a missing client id — the server is not set up.
        log.error("cannot start a MAL link", extra={"error": str(exc)})
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=NOT_CONFIGURED
        ) from exc


@router.get(
    "/callback",
    response_model=None,
    summary="Where MyAnimeList sends the browser back (FR-M1)",
    responses={302: {"description": "redirect to the client's MAL page"}},
)
async def callback(
    user: CurrentUser,
    session: SessionDep,
    settings: SettingsDep,
    code: Annotated[str | None, Query()] = None,
    state: Annotated[str | None, Query()] = None,
    error: Annotated[str | None, Query()] = None,
) -> Response:
    """Complete the handshake and queue the baseline import.

    ``error`` is what MyAnimeList sends when the user presses "Deny"; it is a
    normal outcome, not a failure, and is passed through to the client so it
    can say what happened.
    """
    if error:
        return _back(settings, error=error)
    if not code or not state:
        return _back(settings, error=ERROR_STATE)

    try:
        parsed = oauth.decode_state(settings, state)
    except (oauth.InvalidState, InvalidToken, ConfigurationError) as exc:
        log.info("MAL callback with an unusable state", extra={"error": str(exc)})
        return _back(settings, error=ERROR_STATE)

    if parsed.user_id != user.id:
        # The state was issued to somebody else. Refused rather than honoured:
        # completing it would attach a MyAnimeList account to the wrong Arc
        # user, which is the one mistake this flow must not be able to make.
        log.warning(
            "MAL callback state belongs to another user",
            extra={"session_user": user.id, "state_user": parsed.user_id},
        )
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="state is not yours")

    try:
        async with oauth_client(settings) as client:
            tokens = await client.exchange(code=code, verifier=parsed.verifier)
    except (oauth.MalOAuthError, ConfigurationError) as exc:
        log.warning("MAL token exchange failed", extra={"user_id": user.id, "error": str(exc)})
        return _back(settings, error=ERROR_EXCHANGE)

    link = await _link_of(session, user.id) or MalLink(user_id=user.id)
    store_tokens(settings, link, tokens)
    session.add(link)
    await session.flush()

    # Who this is, on MAL's side. Not fatal if it fails: the tokens are good,
    # the import will work, and a missing username is a cosmetic gap that the
    # next import does not even need to fill.
    try:
        async with client_of(settings, link) as api:
            _, username = await api.me()
        link.mal_username = username
    except MalApiError as exc:
        log.warning(
            "could not read the MAL username",
            extra={"user_id": user.id, "error": str(exc)},
        )

    await enqueue_mal_import(session, user_id=user.id)
    await session.commit()
    log.info("MyAnimeList linked", extra={"user_id": user.id, "mal_username": link.mal_username})
    return _back(settings, linked=True)


@router.delete(
    "/link",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Unlink this account from MyAnimeList",
    responses={404: {"description": NOT_LINKED}},
)
async def unlink(user: CurrentUser, session: SessionDep) -> Response:
    """Delete the link row and its tokens.

    The write log stays, and so do the list entries and their ``mal_dirty``
    flags. The log is an audit trail of things that really happened and is not
    the user's to lose by pressing a button (FR-M5); the dirty flags are the
    truthful statement that Arc holds changes MyAnimeList has not been told
    about, which re-linking should push rather than forget.

    The **queued** rows do not stay queued. There is no longer an account to
    send them to and no job that could: ``mal_push`` would find no link and
    abandon them, and until something did, the sync page would go on counting
    changes as owed that nothing was coming back for. They close ``skipped``
    with the reason, which is the honest reading — nothing was attempted, and
    a decision was made. Rows that had already *failed* are left exactly as
    they are: they record an attempt that really happened, and unlinking is not
    a licence to rewrite it.
    """
    link = await _link_of(session, user.id)
    if link is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=NOT_LINKED)
    closed = await discard_pending(session, user_id=user.id, reason=SKIP_DISCONNECTED)
    await session.delete(link)
    await session.commit()
    log.info("MyAnimeList unlinked", extra={"user_id": user.id, "abandoned": closed})
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post(
    "/import",
    status_code=status.HTTP_202_ACCEPTED,
    summary="Re-import this user's MyAnimeList list now (FR-M2)",
    responses={404: {"description": NOT_LINKED}},
)
async def import_now(user: CurrentUser, session: SessionDep) -> dict[str, int]:
    job = await _queue(session, user_id=user.id, importing=True)
    return {"job_id": job}


@router.post(
    "/push",
    status_code=status.HTTP_202_ACCEPTED,
    summary="Send every pending change to MyAnimeList now (FR-M6)",
    responses={404: {"description": NOT_LINKED}},
)
async def push_now(user: CurrentUser, session: SessionDep) -> dict[str, int]:
    job = await _queue(session, user_id=user.id, importing=False)
    return {"job_id": job}


async def _queue(session: AsyncSession, *, user_id: int, importing: bool) -> int:
    """Queue an import or a full push for a linked user; 404 if there is none."""
    if await session.get(MalLink, user_id) is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=NOT_LINKED)
    job = (
        await enqueue_mal_import(session, user_id=user_id)
        if importing
        else await enqueue_mal_push_all(session, user_id=user_id)
    )
    await session.commit()
    return job.id


@router.get(
    "/log",
    response_model=list[MalWriteOut],
    summary="Every write Arc made to MyAnimeList for this user (FR-M5)",
)
async def write_log(
    user: CurrentUser,
    session: SessionDep,
    limit: Annotated[int, Query(ge=1, le=LOG_MAX_LIMIT)] = LOG_LIMIT,
    status_filter: Annotated[MalWriteStatus | None, Query(alias="status")] = None,
) -> list[MalWriteOut]:
    rows = await writelog.recent(session, user_id=user.id, limit=limit, status=status_filter)
    if not rows:
        return []
    anime_ids = {row.anime_id for row in rows}
    shows = {
        anime.id: anime
        for anime in (await session.scalars(select(Anime).where(Anime.id.in_(anime_ids)))).all()
    }
    # One query for the whole page rather than one existence check per row:
    # ``revertible`` is "this is the newest successful write to its field", and
    # the newest id per (anime, field) answers that for fifty rows at once.
    newest = await writelog.latest_ok_ids(session, user_id=user.id, anime_ids=list(anime_ids))
    # …and one more for the fields with a change still waiting to be sent. A
    # queued row is a newer intention than the newest *successful* write, so
    # offering "revert" on that write would let a click silently supersede a
    # change the user made a moment ago and is still waiting for.
    queued = await writelog.pending_fields(session, user_id=user.id, anime_ids=list(anime_ids))
    statuses = {
        anime_id: entry_status
        for anime_id, entry_status in (
            await session.execute(
                select(ListEntry.anime_id, ListEntry.status).where(
                    ListEntry.user_id == user.id, ListEntry.anime_id.in_(anime_ids)
                )
            )
        ).all()
    }
    return [
        MalWriteOut(
            id=row.id,
            anime=(
                AnimeSummary.from_anime(shows[row.anime_id], statuses.get(row.anime_id))
                if row.anime_id in shows
                else None
            ),
            field=row.field,
            old_value=row.old_value,
            new_value=row.new_value,
            cause=row.cause,
            status=row.status,
            error=row.error,
            created_at=row.created_at,
            revertible=(
                row.status is MalWriteStatus.OK
                and newest.get((row.anime_id, row.field)) == row.id
                and (row.anime_id, row.field) not in queued
            ),
        )
        for row in rows
    ]


@router.post(
    "/log/{log_id}/revert",
    response_model=RevertOut,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Put a field back to the value this write replaced (FR-M5)",
    responses={
        404: {"description": NOT_FOUND},
        409: {"description": f"{NOT_REVERTIBLE}, or {NEWER_QUEUED}"},
    },
)
async def revert(log_id: int, user: CurrentUser, session: SessionDep) -> RevertOut:
    """Write ``old_value`` back locally and queue the push that carries it up.

    A revert is a change like any other: it sets ``updated_by = arc`` and
    ``mal_dirty``, queues a ``mal_push`` with cause ``revert``, and produces
    its own log row when that push runs. So the log always reads as a history
    rather than as a history with holes in it, and reverting the revert is
    simply the newest row becoming revertible in turn.

    Refused while the field has a **queued** row. "This is the newest
    successful write" is not "this is the newest thing the user asked for": a
    change made while the last one was still being retried is newer than any
    of them, and a revert on top of it would replace it without ever saying
    so. The log's ``revertible`` flag is computed the same way, so the button
    is gone before the 409 is needed.
    """
    row = await session.get(MalWriteLog, log_id)
    if row is None or row.user_id != user.id:
        # The same answer for "no such row" and "not yours": a 403 here would
        # tell a caller which log ids exist.
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=NOT_FOUND)
    if await session.get(MalLink, user.id) is None:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=NOT_LINKED)
    if not await writelog.is_revertible(session, row):
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=NOT_REVERTIBLE)
    if await writelog.has_pending_field(
        session, user_id=user.id, anime_id=row.anime_id, field=row.field
    ):
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=NEWER_QUEUED)

    entry = await session.get(ListEntry, (user.id, row.anime_id))
    if entry is None:
        entry = _recreate(user_id=user.id, row=row)
        if entry is None:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=NOT_REVERTIBLE)
        session.add(entry)
        was: Any = None
    else:
        was = _current(entry, row.field)

    try:
        apply_change(entry, field=row.field, value=row.old_value)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc

    entry.updated_by = UpdatedBy.ARC
    entry.mal_dirty = True
    entry.updated_at = datetime.now(UTC)
    await session.flush()
    # The third of FR-M7's user-originated events, and the third and last
    # caller allowed to queue a write. The row carries ``revert`` so the push
    # knows that lowering progress or clearing a score is exactly what was
    # asked for — a guard that applies to an automatic event does not apply
    # to a button the user pressed.
    await writelog.record_pending(
        session,
        user_id=user.id,
        anime_id=row.anime_id,
        field=row.field,
        old_value=was,
        new_value=row.old_value,
        cause=MalWriteCause.REVERT,
    )
    job = await enqueue_mal_push(session, user_id=user.id, anime_id=row.anime_id)
    await session.commit()
    return RevertOut(
        anime_id=row.anime_id,
        field=row.field,
        value=row.old_value,
        job_id=job.id if job else None,
    )


def _current(entry: ListEntry, field: str) -> Any:
    """What the entry holds for one field name, before the revert replaces it."""
    if field == writelog.FIELD_STATUS:
        return entry.status.value
    if field == writelog.FIELD_SCORE:
        return entry.score
    return entry.progress


def _recreate(*, user_id: int, row: MalWriteLog) -> ListEntry | None:
    """The entry a reverted *removal* puts back, or ``None`` if it cannot.

    Only a status row with a real ``old_value`` describes a whole entry. A
    score or a progress row on a show that is no longer on the list says
    nothing about what status it should come back as, and inventing one would
    be Arc making a claim the user did not.
    """
    if row.field != writelog.FIELD_STATUS or row.old_value is None:
        return None
    try:
        return ListEntry(
            user_id=user_id,
            anime_id=row.anime_id,
            status=ListStatus(str(row.old_value)),
            progress=0,
        )
    except ValueError:  # pragma: no cover - the column only ever held a status
        return None


def _back(settings: Settings, *, linked: bool = False, error: str | None = None) -> Response:
    """Send the browser back to the client's MAL page with the outcome.

    The error is url-encoded because it is not always one of the two codes
    above: MyAnimeList's own ``?error=`` comes back verbatim when the user
    presses "Deny", and a value with an ``&`` or a space in it would otherwise
    become a second query parameter — or a header nobody meant to send.
    """
    query = "linked=1" if linked else urlencode({"error": error or ERROR_STATE})
    return RedirectResponse(
        url=f"{settings.public_url.rstrip('/')}{CLIENT_PATH}?{query}",
        status_code=status.HTTP_302_FOUND,
    )


__all__ = [
    "ALREADY_LINKED",
    "CLIENT_PATH",
    "ERROR_EXCHANGE",
    "ERROR_STATE",
    "LOG_LIMIT",
    "LOG_MAX_LIMIT",
    "NEWER_QUEUED",
    "NOT_CONFIGURED",
    "NOT_FOUND",
    "NOT_LINKED",
    "NOT_REVERTIBLE",
    "MalLinkOut",
    "MalStatusOut",
    "MalSyncOut",
    "MalWriteOut",
    "RevertOut",
    "router",
]
